tests/e2e/slashing.go

```diff
@@ -374,106 +375,371 @@ func (suite *CCVTestSuite) TestOnRecvSlashPacketErrors() {
+				relayAllCommittedPackets(
+					suite,
+					suite.consumerChain,
+					suite.path,
+					ccv.ConsumerPortID,
+					suite.path.EndpointA.ChannelID,
+					1,
+				)
+			},
+			func(delAddr sdk.AccAddress, initBalance sdk.Int) {
+				// the undelegation was only matured on the consumer,
+				// thus the slash was successful
+				expectedBalance := initBalance.Sub(halfBondAmt).Sub(slashAmount)
+				suite.Require().Equal(
+					expectedBalance,
+					getBalance(suite, suite.providerCtx(), delAddr),
+					"delegator should have been slashed",
+				)
 			},
-		}, {
-			func(suite *CCVTestSuite) error {
-				return nil
+		},
+		// delegate - infraction - undelegate - mature consumer - mature provider - slash
+		// no slash
+		{
+			"infraction before undelegate, detected after maturity on both chain",
+			func(consAddr sdk.ConsAddress) {
+				// increment time so that the unbonding period ends on the consumer
+				incrementTime(suite, consumerUnbondingPeriod+time.Hour)
+
+				// relay all packets from consumer to provider
+				relayAllCommittedPackets(
+					suite,
+					suite.consumerChain,
+					suite.path,
+					ccv.ConsumerPortID,
+					suite.path.EndpointA.ChannelID,
+					2,
+				)
+
+				// increment time so that the unbonding period ends on the provider
+				incrementTime(suite, providerUnbondingPeriod)
+
+				// slash
+				suite.consumerApp.GetConsumerKeeper().Slash(
+					suite.consumerCtx(),
+					consAddr,
+					int64(undelegateConsumerHeight)-1,
+					powerBeforeUndelegate,
+					sdk.ZeroDec(),
+					stakingtypes.Downtime,
+				)
+
+				// commit state on consumer
+				suite.coordinator.CommitBlock(suite.consumerChain)
+
+				// relay all packets from consumer to provider
+				relayAllCommittedPackets(
+					suite,
+					suite.consumerChain,
+					suite.path,
+					ccv.ConsumerPortID,
+					suite.path.EndpointA.ChannelID,
+					1,
+				)
 			},
-		}, {
-			// undelegate a quarter of the new shares created
-			func(suite *CCVTestSuite) error {
-				_, err = providerStakingKeeper.Undelegate(suite.providerChain.GetContext(), delAddr, valAddr, testShares.QuoInt64(4))
-				return err
+			func(delAddr sdk.AccAddress, initBalance sdk.Int) {
+				// the undelegation is already matured, thus it is not slashed
+				expectedBalance := initBalance.Sub(halfBondAmt)
+				suite.Require().Equal(
+					expectedBalance,
+					getBalance(suite, suite.providerCtx(), delAddr),
+					"delegator shouldn't have been slashed",
+				)
 			},
 		},
 	}
 
-	// execute the setup operations, distributed uniformly in three blocks.
-	// For each of them, save their current VSC Id value which map correspond respectively
-	// to the block heights lesser, equal and greater than the undelegation creation height.
-	vscIDs := make([]uint64, 0, 3)
-	for _, so := range setupOperations {
-		err := so.fn(suite)
-		suite.Require().NoError(err)
+	for i, tc := range testCases {
+		providerKeeper := suite.providerApp.GetProviderKeeper()
```

Comment:
To fully reset the state, we need to call SetupTest on e2e tests that have multiple cases. Otherwise I believe we're using the same store between cases
```suggestion
                suite.SetupTest()
		providerKeeper := suite.providerApp.GetProviderKeeper()
```

Reply:
Already doing it at the end of the for loop.